util/cgroup/cgroup.go

```diff
@@ -0,0 +1,411 @@
+	return undo, nil
+}
+
+func readFile(filepath string) (res []byte, err error) {
+	var f *os.File
+	//nolint:gosec
+	f, err = os.Open(filepath)
+	if err != nil {
+		return nil, err
+	}
+	defer func() {
+		err = errors.CombineErrors(err, f.Close())
+	}()
+	res, err = io.ReadAll(f)
+	return res, err
+}
+
+// The controller is defined via either type `memory` for cgroup v1 or via empty type for cgroup v2,
+// where the type is the second field in /proc/[pid]/cgroup file
+func detectControlPath(cgroupFilePath string, controller string) (string, error) {
+	//nolint:gosec
+	cgroup, err := os.Open(cgroupFilePath)
+	if err != nil {
+		return "", errors.Wrapf(err, "failed to read %s cgroup from cgroups file: %s", controller, cgroupFilePath)
+	}
+	defer func() {
+		err := cgroup.Close()
+		if err != nil {
+			log.Error("close cgroupFilePath", zap.Error(err))
+		}
+	}()
+
+	scanner := bufio.NewScanner(cgroup)
+	var unifiedPathIfFound string
+	for scanner.Scan() {
+		fields := bytes.Split(scanner.Bytes(), []byte{':'})
+		if len(fields) != 3 {
+			// The lines should always have three fields, there's something fishy here.
+			continue
+		}
+
+		f0, f1 := string(fields[0]), string(fields[1])
+		// First case if v2, second - v1. We give v2 the priority here.
+		// There is also a `hybrid` mode when both  versions are enabled,
+		// but no known container solutions support it.
+		if f0 == "0" && f1 == "" {
+			unifiedPathIfFound = string(fields[2])
+		} else if f1 == controller {
+			return string(fields[2]), nil
+		}
+	}
+
+	return unifiedPathIfFound, nil
+}
+
+// See http://man7.org/linux/man-pages/man5/proc.5.html for `mountinfo` format.
+func getCgroupDetails(mountInfoPath string, cRoot string, controller string) (string, int, error) {
+	//nolint:gosec
+	info, err := os.Open(mountInfoPath)
+	if err != nil {
+		return "", 0, errors.Wrapf(err, "failed to read mounts info from file: %s", mountInfoPath)
+	}
+	defer func() {
+		err := info.Close()
+		if err != nil {
+			log.Error("close mountInfoPath", zap.Error(err))
+		}
+	}()
+
+	scanner := bufio.NewScanner(info)
+	for scanner.Scan() {
+		fields := bytes.Fields(scanner.Bytes())
+		if len(fields) < 10 {
+			continue
+		}
+
+		ver, ok := detectCgroupVersion(fields, controller)
+		if ok {
+			mountPoint := string(fields[4])
+			if ver == 2 {
+				return mountPoint, ver, nil
+			}
+			// It is possible that the controller mount and the cgroup path are not the same (both are relative to the NS root).
+			// So start with the mount and construct the relative path of the cgroup.
+			// To test:
+			//   cgcreate -t $USER:$USER -a $USER:$USER -g memory:crdb_test
+			//   echo 3999997952 > /sys/fs/cgroup/memory/crdb_test/memory.limit_in_bytes
+			//   cgexec -g memory:crdb_test ./cockroach start-single-node
+			// cockroach.log -> server/config.go:433 ⋮ system total memory: ‹3.7 GiB›
+			//   without constructing the relative path
+			// cockroach.log -> server/config.go:433 ⋮ system total memory: ‹63 GiB›
```

Comment:
Should we keep `cockroach` in comments? not sure about it.

Reply:
I updated the content.